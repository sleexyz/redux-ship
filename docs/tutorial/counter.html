
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>Counter Â· Redux Ship</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.2">
        
        
        
    
    <link rel="stylesheet" href="../../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="http-request.html" />
    
    
    <link rel="prev" href="./" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../../">
            
                <a href="../../">
            
                    
                    Read Me
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="./">
            
                <a href="./">
            
                    
                    Tutorial
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter active" data-level="1.2.1" data-path="counter.html">
            
                <a href="counter.html">
            
                    
                    Counter
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.2" data-path="http-request.html">
            
                <a href="http-request.html">
            
                    
                    HTTP Request
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.3" data-path="composition.html">
            
                <a href="composition.html">
            
                    
                    Composition
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.4" data-path="testing.html">
            
                <a href="testing.html">
            
                    
                    Testing
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../api.html">
            
                <a href="../api.html">
            
                    
                    API Reference
            
                </a>
            

            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href="../.." >Counter</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="counter">Counter</h1>
<p>We start by the simple example of a counter. A counter does not have asynchronous actions so Redux Ship is not necessary here, but it helps to get started.</p>
<p><img src="https://raw.githubusercontent.com/clarus/redux-ship/master/docs/tutorial/counter.png" alt="Screenshot" width="150px"></p>
<ul>
<li><a href="#create-a-react-application">Create a React application</a></li>
<li><a href="#model">Model</a></li>
<li><a href="#view">View</a></li>
<li><a href="#controller">Controller</a></li>
<li><a href="#wrapping-everything-up">Wrapping everything up</a></li>
<li><a href="#snapshots">Snapshots</a></li>
</ul>
<h2 id="create-a-react-application">Create a React application</h2>
<p>Use <a href="https://github.com/facebookincubator/create-react-app#getting-started" target="_blank"><code>create-react-app</code></a> to create a <code>counter</code> application:</p>
<pre><code>create-react-app counter
</code></pre><p>You should have the following files in <code>src/</code>:</p>
<pre><code>App.css
App.js
App.test.js
index.css
index.js
logo.svg
</code></pre><p>Install <a href="http://redux.js.org/" target="_blank">Redux</a> and Redux Ship with their dev tools:</p>
<pre><code>npm install --save redux redux-ship redux-logger redux-ship-logger babel-polyfill
</code></pre><p>Optionally setup the <a href="https://flowtype.org/" target="_blank">Flow</a> type system:</p>
<pre><code>npm install -g flow-bin
flow init
flow
</code></pre><h2 id="model">Model</h2>
<p>We add a <code>model.js</code> file to describe the &quot;Redux part&quot; of the counter:</p>
<pre><code class="lang-js"><span class="hljs-comment">// @flow</span>
<span class="hljs-keyword">export</span> type State = number;

<span class="hljs-keyword">export</span> <span class="hljs-keyword">const</span> initialState = <span class="hljs-number">0</span>;

<span class="hljs-keyword">export</span> type Commit = {
  type: <span class="hljs-string">&apos;Increment&apos;</span>,
} | {
  type: <span class="hljs-string">&apos;Decrement&apos;</span>,
};

<span class="hljs-keyword">export</span> <span class="hljs-function"><span class="hljs-keyword">function</span> <span class="hljs-title">reduce</span>(<span class="hljs-params">state: State, commit: Commit</span>): <span class="hljs-title">State</span> </span>{
  <span class="hljs-keyword">switch</span> (commit.type) {
  <span class="hljs-keyword">case</span> <span class="hljs-string">&apos;Increment&apos;</span>:
    <span class="hljs-keyword">return</span> state + <span class="hljs-number">1</span>;
  <span class="hljs-keyword">case</span> <span class="hljs-string">&apos;Decrement&apos;</span>:
    <span class="hljs-keyword">return</span> state - <span class="hljs-number">1</span>;
  <span class="hljs-keyword">default</span>:
    <span class="hljs-keyword">return</span> state;
  }
}
</code></pre>
<p>This describes a quite standard reducer with typing. Notice that we name the actions <code>Commit</code>. This is to make clear that these actions are used to modify the state, by opposition to asynchronous actions.</p>
<h2 id="view">View</h2>
<p>In <code>App.js</code> we add the view of the counter:</p>
<pre><code class="lang-js"><span class="hljs-comment">// @flow</span>
<span class="hljs-keyword">import</span> React, { PureComponent } <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;react&apos;</span>;
<span class="hljs-keyword">import</span> <span class="hljs-string">&apos;./App.css&apos;</span>;
<span class="hljs-keyword">import</span> * <span class="hljs-keyword">as</span> Controller <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;./controller&apos;</span>;
<span class="hljs-keyword">import</span> * <span class="hljs-keyword">as</span> Model <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;./model&apos;</span>;

type Props = {
  dispatch: (action: Controller.Action) =&gt; <span class="hljs-keyword">void</span>,
  state: Model.State,
};

<span class="hljs-keyword">export</span> <span class="hljs-keyword">default</span> <span class="hljs-class"><span class="hljs-keyword">class</span> <span class="hljs-title">App</span> <span class="hljs-keyword">extends</span> <span class="hljs-title">PureComponent</span>&lt;<span class="hljs-title">void</span>, <span class="hljs-title">Props</span>, <span class="hljs-title">void</span>&gt; </span>{
  handleClickIncrement = (): <span class="hljs-keyword">void</span> =&gt; {
    <span class="hljs-keyword">this</span>.props.dispatch({type: <span class="hljs-string">&apos;ClickIncrement&apos;</span>});
  };

  handleClickDecrement = (): <span class="hljs-keyword">void</span> =&gt; {
    <span class="hljs-keyword">this</span>.props.dispatch({type: <span class="hljs-string">&apos;ClickDecrement&apos;</span>});
  };

  render() {
    <span class="hljs-keyword">return</span> (
      <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-name">div</span> <span class="hljs-attr">className</span>=<span class="hljs-string">&quot;App&quot;</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-name">p</span>&gt;</span>{this.props.state}<span class="hljs-tag">&lt;/<span class="hljs-name">p</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-name">button</span> <span class="hljs-attr">onClick</span>=<span class="hljs-string">{this.handleClickIncrement}</span>&gt;</span>
          +1
        <span class="hljs-tag">&lt;/<span class="hljs-name">button</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-name">button</span> <span class="hljs-attr">onClick</span>=<span class="hljs-string">{this.handleClickDecrement}</span>&gt;</span>
          -1
        <span class="hljs-tag">&lt;/<span class="hljs-name">button</span>&gt;</span>
      <span class="hljs-tag">&lt;/<span class="hljs-name">div</span>&gt;</span></span>
    );
  }
}
</code></pre>
<p>We display the value of the counter with <code>&lt;p&gt;{this.props.state}&lt;/p&gt;</code>. To handle the clicks on the buttons <code>+1</code> and <code>-1</code> we dispatch actions to the controller.</p>
<h2 id="controller">Controller</h2>
<p>We define the controller in <code>controller.js</code>:</p>
<pre><code class="lang-js"><span class="hljs-comment">// @flow</span>
<span class="hljs-keyword">import</span> * <span class="hljs-keyword">as</span> Ship <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;redux-ship&apos;</span>;
<span class="hljs-keyword">import</span> * <span class="hljs-keyword">as</span> Model <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;./model&apos;</span>;

<span class="hljs-keyword">export</span> type Action = {
  type: <span class="hljs-string">&apos;ClickIncrement&apos;</span>,
} | {
  type: <span class="hljs-string">&apos;ClickDecrement&apos;</span>,
};

<span class="hljs-keyword">export</span> <span class="hljs-function"><span class="hljs-keyword">function</span>* <span class="hljs-title">control</span>(<span class="hljs-params">action: Action</span>): <span class="hljs-title">Ship</span>.<span class="hljs-title">Ship</span>&lt;*, <span class="hljs-title">Model</span>.<span class="hljs-title">Commit</span>, <span class="hljs-title">Model</span>.<span class="hljs-title">State</span>, <span class="hljs-title">void</span>&gt; </span>{
  <span class="hljs-keyword">switch</span> (action.type) {
  <span class="hljs-keyword">case</span> <span class="hljs-string">&apos;ClickIncrement&apos;</span>:
    <span class="hljs-keyword">yield</span>* Ship.commit({type: <span class="hljs-string">&apos;Increment&apos;</span>});
    <span class="hljs-keyword">return</span>;
  <span class="hljs-keyword">case</span> <span class="hljs-string">&apos;ClickDecrement&apos;</span>:
    <span class="hljs-keyword">yield</span>* Ship.commit({type: <span class="hljs-string">&apos;Decrement&apos;</span>});
    <span class="hljs-keyword">return</span>;
  <span class="hljs-keyword">default</span>:
    <span class="hljs-keyword">return</span>;
  }
}
</code></pre>
<p>The controller describes how to react to the application events, here a click on <code>+1</code> or <code>-1</code>. We define a type <code>Action</code> which is the type of all these application events. The function <code>control</code> handles an action by returning a ship.</p>
<p>A ship is the description of a side effect, including API calls, modifications of the Redux state, timers, url update, or calls to third-party libraries with side effects. By side effect we mean &quot;anything which is not purely functional&quot;. We define a ship with a generator and Redux Ship primitives.</p>
<p>We call:</p>
<pre><code class="lang-js"><span class="hljs-keyword">yield</span>* Ship.commit({type: <span class="hljs-string">&apos;Increment&apos;</span>});
</code></pre>
<p>to commit a commit to the Redux state. All functions in Redux Ship are called with <code>yield*</code>, you should never encounter a <code>yield</code>. We avoid the <code>yield</code> operator because it is a difficult to type in Flow, and instead call proxy functions with <code>yield*</code>.</p>
<p>The return type of <code>control</code> is:</p>
<pre><code class="lang-js">Ship.Ship&lt;*, Model.Commit, Model.State, <span class="hljs-keyword">void</span>&gt;
</code></pre>
<p>meaning that this controller is attached to our model. We could not for example run:</p>
<pre><code class="lang-js"><span class="hljs-keyword">yield</span>* Ship.commit({type: <span class="hljs-string">&apos;Foo&apos;</span>}); <span class="hljs-comment">// error</span>
</code></pre>
<p>as this would result in a Flow type error since <code>{type: &apos;Foo&apos;}</code> is not of type <code>Model.Commit</code>.</p>
<h2 id="wrapping-everything-up">Wrapping everything up</h2>
<p>We instantiate a Redux store together with the Redux Ship middleware in <code>store.js</code>:</p>
<pre><code class="lang-js"><span class="hljs-comment">// @flow</span>
<span class="hljs-keyword">import</span> {applyMiddleware, createStore} <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;redux&apos;</span>;
<span class="hljs-keyword">import</span> * <span class="hljs-keyword">as</span> Ship <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;redux-ship&apos;</span>;
<span class="hljs-keyword">import</span> createLogger <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;redux-logger&apos;</span>;
<span class="hljs-keyword">import</span> {logControl} <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;redux-ship-logger&apos;</span>;
<span class="hljs-keyword">import</span> * <span class="hljs-keyword">as</span> Controller <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;./controller&apos;</span>;
<span class="hljs-keyword">import</span> * <span class="hljs-keyword">as</span> Model <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;./model&apos;</span>;

<span class="hljs-function"><span class="hljs-keyword">function</span> <span class="hljs-title">runEffect</span>(<span class="hljs-params"></span>) </span>{}

<span class="hljs-keyword">const</span> middlewares = [
  Ship.middleware(runEffect, logControl(Controller.control)),
  createLogger(),
];

<span class="hljs-keyword">export</span> <span class="hljs-keyword">default</span> createStore(
  Model.reduce,
  Model.initialState,
  applyMiddleware(...middlewares)
);
</code></pre>
<p>We provide two parameters to the <code>Ship.middleware</code> function:</p>
<ul>
<li><code>runEffect</code> which is empty for now, as we only have synchronous actions;</li>
<li><code>logControl(Controller.control))</code> which is our controller. We wrap it with a <code>logControl</code> to add loggin to the Redux Ship actions.</li>
</ul>
<p>We bootstrap the application in <code>index.js</code>:</p>
<pre><code class="lang-js"><span class="hljs-comment">// @flow</span>
<span class="hljs-keyword">import</span> <span class="hljs-string">&apos;babel-polyfill&apos;</span>;
<span class="hljs-keyword">import</span> React <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;react&apos;</span>;
<span class="hljs-keyword">import</span> ReactDOM <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;react-dom&apos;</span>;
<span class="hljs-keyword">import</span> App <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;./App&apos;</span>;
<span class="hljs-keyword">import</span> <span class="hljs-string">&apos;./index.css&apos;</span>;
<span class="hljs-keyword">import</span> store <span class="hljs-keyword">from</span> <span class="hljs-string">&apos;./store&apos;</span>;

<span class="hljs-function"><span class="hljs-keyword">function</span> <span class="hljs-title">render</span>(<span class="hljs-params"></span>): <span class="hljs-title">void</span> </span>{
  ReactDOM.render(
    <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-name">App</span> <span class="hljs-attr">dispatch</span>=<span class="hljs-string">{store.dispatch}</span> <span class="hljs-attr">state</span>=<span class="hljs-string">{store.getState()}</span> /&gt;</span>,
    document.getElementById(&apos;root&apos;)
  );
}

store.subscribe(render);
render();
</span></code></pre>
<p>We define a <a href="https://facebook.github.io/react/" target="_blank">React</a> <code>render</code> function to render the application. We subscribe to the <code>store</code> to re-render when the Redux store is updated. Note that you can also use <a href="https://github.com/reactjs/react-redux" target="_blank">react-redux</a> to connect React to Redux.</p>
<h2 id="snapshots">Snapshots</h2>
<p>When we look at our browser&apos;s console we see something like:</p>
<p><img src="https://raw.githubusercontent.com/clarus/redux-ship/master/docs/tutorial/counter-logs.png" alt="Logs" width="600px"></p>
<p>We have an action <code>{type: &apos;Increment&apos;}</code> which takes us from the state <code>0</code> to the state <code>1</code>. This action is logged by <a href="https://github.com/evgenyrodionov/redux-logger" target="_blank">redux-logger</a>. We also have a line:</p>
<pre><code>control @ 19:35:41.214 ClickIncrement
</code></pre><p>which is the log of our controller as given by <a href="https://github.com/clarus/redux-ship-logger" target="_blank">redux-ship-logger</a>. We see the snapshot of our controller:</p>
<pre><code class="lang-js">[
  {
    type: <span class="hljs-string">&apos;Commit&apos;</span>,
    commit: {type: <span class="hljs-string">&apos;Increment&apos;</span>}
  }
]
</code></pre>
<p>which is an array of one element, the commit of <code>{type: &apos;Increment&apos;}</code>, describing all what the controller has done. Nothing fancy there, but the snapshots of our controllers will become increasingly useful as we design more complex controllers.</p>
<p>Let us move to the <a href="http-request.html">HTTP Request</a> section to see how to make asynchronous actions.</p>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="./" class="navigation navigation-prev " aria-label="Previous page: Tutorial">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
                <a href="http-request.html" class="navigation navigation-next " aria-label="Next page: HTTP Request">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"Counter","level":"1.2.1","depth":2,"next":{"title":"HTTP Request","level":"1.2.2","depth":2,"path":"docs/tutorial/http-request.md","ref":"docs/tutorial/http-request.md","articles":[]},"previous":{"title":"Tutorial","level":"1.2","depth":1,"path":"docs/tutorial/README.md","ref":"docs/tutorial/README.md","articles":[{"title":"Counter","level":"1.2.1","depth":2,"path":"docs/tutorial/counter.md","ref":"docs/tutorial/counter.md","articles":[]},{"title":"HTTP Request","level":"1.2.2","depth":2,"path":"docs/tutorial/http-request.md","ref":"docs/tutorial/http-request.md","articles":[]},{"title":"Composition","level":"1.2.3","depth":2,"path":"docs/tutorial/composition.md","ref":"docs/tutorial/composition.md","articles":[]},{"title":"Testing","level":"1.2.4","depth":2,"path":"docs/tutorial/testing.md","ref":"docs/tutorial/testing.md","articles":[]}]},"dir":"ltr"},"config":{"plugins":["github"],"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"pluginsConfig":{"github":{"url":"https://github.com/clarus/redux-ship"},"highlight":{},"search":{},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"fontsettings":{"theme":"white","family":"sans","size":2},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"theme":"default","pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"variables":{},"title":"Redux Ship","gitbook":"*"},"file":{"path":"docs/tutorial/counter.md","mtime":"2016-12-09T13:03:51.702Z","type":"markdown"},"gitbook":{"version":"3.2.2","time":"2016-12-09T13:05:11.921Z"},"basePath":"../..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../../gitbook/gitbook.js"></script>
    <script src="../../gitbook/theme.js"></script>
    
        
        <script src="../../gitbook/gitbook-plugin-github/plugin.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

