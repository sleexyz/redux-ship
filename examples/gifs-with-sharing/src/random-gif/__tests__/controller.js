// @flow
import * as Test from '../../test';
import * as Effect from '../../effect';
import * as RandomGifController from '../controller';
import * as CounterModel from '../../counter/model';
import * as RandomGifModel from '../model';

const defaultState = {
  counter: CounterModel.initialState,
  randomGif: RandomGifModel.initialState,
};

Test.snapshotApplyCommit(RandomGifController.applyCommit, {
  loadStart: {
    commit: {type: 'LoadStart'},
    state: defaultState,
  },
  loadSuccessIncrementByOne: {
    commit: {
      type: 'LoadSuccess',
      gifUrl: 'https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif',
    },
    state: defaultState,
  },
  loadSuccessIncrementByTwo: {
    commit: {
      type: 'LoadSuccess',
      gifUrl: 'https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif',
    },
    state: {
      ...defaultState,
      counter: {
        ...defaultState.counter,
        count: 10,
      },
    },
  },
});

function reduce(
  state: RandomGifController.State,
  patch: RandomGifController.Patch
): RandomGifController.State {
  return {
    ...state,
    ...patch.counter &&
      {counter: CounterModel.reduce(state.counter, patch.counter)},
    ...patch.randomGif &&
      {randomGif: RandomGifModel.reduce(state.randomGif, patch.randomGif)},
  };
}

function runEffect(effect: Effect.Effect): any {
  switch (effect.type) {
  case 'HttpRequest':
    return JSON.stringify({
      data: {
        image_url: 'https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif',
      },
    });
  default:
    return;
  }
}

Test.snapshotControl(
  RandomGifController.control,
  RandomGifController.applyCommit,
  reduce,
  {
    load: {
      action: {
        type: 'Load',
        tag: 'cats',
      },
      runEffect,
      state: defaultState,
    },
  }
);
