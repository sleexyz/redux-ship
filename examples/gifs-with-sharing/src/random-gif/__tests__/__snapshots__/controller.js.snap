exports[`test load 1`] = `
Array [
  Object {
    "commit": Object {
      "type": "LoadStart"
    },
    "type": "Commit"
  },
  Object {
    "effect": Object {
      "type": "HttpRequest",
      "url": "http://api.giphy.com/v1/gifs/random?api_key=dc6zaTOxFJmzC&tag=cats"
    },
    "result": "{"data":{"image_url":"https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif"}}",
    "type": "Effect"
  },
  Object {
    "commit": Object {
      "gifUrl": "https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif",
      "type": "LoadSuccess"
    },
    "type": "Commit"
  }
]
`;

exports[`test loadStart 1`] = `
Object {
  "randomGif": Object {
    "type": "LoadStart"
  }
}
`;

exports[`test loadSuccessIncrementByOne 1`] = `
Object {
  "counter": Object {
    "type": "IncrementByOne"
  },
  "randomGif": Object {
    "gifUrl": "https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif",
    "type": "LoadSuccess"
  }
}
`;

exports[`test loadSuccessIncrementByTwo 1`] = `
Object {
  "counter": Object {
    "type": "IncrementByTwo"
  },
  "randomGif": Object {
    "gifUrl": "https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif",
    "type": "LoadSuccess"
  }
}
`;
